common/functions-v2/src/scalars/string.rs

```diff
@@ -156,3 +289,48 @@ fn vectorize_string_to_string(
         }
     }
 }
+
+// Vectorize (string, string) -> string function with customer estimate_bytes.
+// This function is only used for trim_[leading|trailing|both] function.
+// When arg1 is a scalar, arg2 must be a scalar.
+fn vectorize_two_string_to_string_from_trim(
+    estimate_bytes: impl Fn(&StringColumn) -> usize + Copy,
+    func: impl Fn(&[u8], &[u8], &mut StringColumnBuilder) -> Result<(), String> + Copy,
+) -> impl Fn(
+    ValueRef<StringType>,
+    ValueRef<StringType>,
+    &GenericMap,
+) -> Result<Value<StringType>, String>
++ Copy {
+    move |arg1, arg2, _| match (arg1, arg2) {
+        (ValueRef::Scalar(val), ValueRef::Scalar(trim_str)) => {
+            let mut builder = StringColumnBuilder::with_capacity(1, 0);
+            func(val, trim_str, &mut builder)?;
+            Ok(Value::Scalar(builder.build_scalar()))
+        }
+        (ValueRef::Column(col), ValueRef::Scalar(trim_str)) => {
+            let data_capacity = estimate_bytes(&col);
+            let mut builder = StringColumnBuilder::with_capacity(col.len(), data_capacity);
+            for val in col.iter() {
+                func(val, trim_str, &mut builder)?;
+            }
+            Ok(Value::Column(builder.build()))
+        }
+        (ValueRef::Column(col), ValueRef::Column(trim_str)) => {
+            let data_capacity = estimate_bytes(&col);
+            let mut builder = StringColumnBuilder::with_capacity(col.len(), data_capacity);
+            for (idx, val) in col.iter().enumerate() {
+                if let Some(trim_val) = trim_str.index(idx) {
+                    func(val, trim_val, &mut builder)?;
+                } else {
+                    return Err(format!("Column length mismatch at index {}", idx));
+                }
+            }
+            Ok(Value::Column(builder.build()))
+        }
+        (_, _) => {
+            // Unreacachable. If arg1 is a scalar, arg2 must be a scalar.
```

Comment:
why?

Reply:
According to the behavior in old expression framework:

```
mysql> set enable_planner_v2 =1;
Query OK, 0 rows affected (0.04 sec)

mysql> select trim_leading('a', b) from test.t;
ERROR 1105 (HY000): Code: 1058, displayText = downcast column error, column type: "Const(String)", expected column: "common_datavalues::columns::string::StringColumn" (while in processor thread 0).
```

Reply:
That's a bug!

Reply:
😂.